rule make_up_results:
	output: "the_results.txt"
	shell: r"sleep 1m; touch the_results.txt"

rule trim_reads:
    input: 
        read_1 = "../input_data/cage_data/Fastq_R1/{sample}_R1.fastq.gz",
        read_2 = "../input_data/cage_data/Fastq_R2/{sample}_R2.fastq.gz",
        solexa = "../input_data/solexa_sequencing_primers.fa",
    threads: 2 #Â Adjust later or maybe put in config file?
    log: "../intermediate_data/snakemake_intermediate_data/logs/{sample}.trim_reads.log"
    shell: 
        r"""
        source package 50fcf79b-73a3-4f94-9553-5ed917823423
        trimmomatic PE -threads {threads}  {input.read_1} {input.read_2} -baseout ${sample} ILLUMINACLIP:../{input.solexa}:2:30:12 SLIDINGWINDOW:4:20 MINLEN:20 AVGQUAL:20 
        """


rule fastqc:
    input:  "../intermediate_data/snakemake_intermediate_data/{myfile}.fq.gz"
    output:
        html="../intermediate_data/snakemake_intermediate_data/{myfile}_fastqc.html",
        zip="../intermediate_data/snakemake_intermediate_data/{myfile}_fastqc.zip"
    threads: 2 # adjust as needed
    log: "../intermediate_data/snakemake_intermediate_data/logs/fastqc.log"
    shell:
        """
        # Define memory per thread
        mem_mb=$(python -c 'print({{int({resources.mem_mb} / {threads})}})')

        # Function to get basename without extension
        basename_without_ext() {{
            base=$(basename "$1")
            base=${{base//.gz/}}
            base=${{base//.fastq/}}
            base=${{base//.fq/}}
            echo "$base"
        }}
        # Create temporary directory
        tempdir=$(mktemp -d)
        trap "rm -rf $tempdir" EXIT

        fastqc --threads {threads} --memory $mem_mb {params.extra} --outdir $tempdir {input} &> {log}

        # Move outputs to proper location
        output_base=$(basename_without_ext {input})
        html_path="$tempdir/${{output_base}}_fastqc.html"
        zip_path="$tempdir/${{output_base}}_fastqc.zip"
        mv $html_path {output.html}
        mv $zip_path {output.zip}
        """









rule star_index:
    output: directory("../intermediate_data/snakemake_intermediate_data/fielder_star_index") # Output is a directory 
    input:
        reference_fasta = "../../input_data/fielder_genome_data/201215_Fielder_pseudomolecules_V1+unanchored_contigs.fasta",
        reference_gtf = "../../input_data/fielder_genome_data/fielder.release.gtf"
    threads: 16
    shell: r""" source package /tgac/software/testing/bin/STAR-2.3.0 # using version 2.3.0 as later versions seem to have bugs
    STAR --runThreadN {threads} --runMode genomeGenerate --genomeDir {output} --genomeFastaFiles {input.reference_fasta} \
     --sjdbGTFfile {input.reference_gtf} --sjdbOverhang 100 --limitGenomeGenerateRAM 100111047092
    """




# rule bowtie_index:
#     output: "../intermediate_data/snakemake_intermediate_data/"
# rule bwa_index:
#     output: "../intermediate_data/snakemake_intermediate_data/"


# rule hisat_index:
#     output: "../intermediate_data/snakemake_intermediate_data/hisat_fielder.idx"
#     shell:r""" source package f9c1e0c5-d0e8-4ba0-9edd-88235400fa13
#         hisat2_extract_splice_sites.py fielder.release.gtf > fielder.ss
#         hisat2_extract_exons.py fielder.release.gtf > fielder.exon
#         hisat2-build -p 16 --ss fielder.ss --exon fielder.exon 201216_Fielder_pseudomolecules_V1+unanchored_contigs.fasta fielder.idx
#     """
#     log: "../intermediate_data/snakemake_intermediate_data/hisat.log"

# rule hisat_align:
#     output: "../intermediate_data/snakemake_intermediate_data/{sample}.hisat.bam"
#     input: 
#         index = "../intermediate_data/snakemake_intermediate_data/hisat_fielder.idx", 
#         read_1 = "../intermediate_data/snakemake_intermediate_data/{sample}.trimmed.fastq.gz"
#         read_2 =  "../intermediate_data/snakemake_intermediate_data/{sample}.trimmed.fastq.gz"
#     shell: r"""
#     source package f9c1e0c5-d0e8-4ba0-9edd-88235400fa13 # source hisat2 
#     source package 638df626-d658-40aa-80e5-14a275b7464b # source samtools 
#     hisat2 -x f{input.index} -p 16 -1 {input.read_1} -2 {input.read_1} | \
#     samtools view -bS > ../intermediate_data/snakemake_intermediate_data/{sample}.hisat.bam
#     """"

